homeassistant/components/volvooncall/__init__.py

```diff
@@ -128,76 +129,48 @@ async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
 
-    def is_enabled(attr):
-        """Return true if the user has enabled the resource."""
-        return attr in config[DOMAIN].get(CONF_RESOURCES, [attr])
+    data = hass.data[DATA_KEY] = VolvoData(hass, connection, config, coordinator)
 
-    def discover_vehicle(vehicle):
-        """Load relevant platforms."""
-        data.vehicles.add(vehicle.vin)
+    _LOGGER.info("Logging in to service")
+    return await data.update()
 
-        dashboard = vehicle.dashboard(
-            mutable=config[DOMAIN][CONF_MUTABLE],
-            scandinavian_miles=config[DOMAIN][CONF_SCANDINAVIAN_MILES],
-        )
 
-        for instrument in (
-            instrument
-            for instrument in dashboard.instruments
-            if instrument.component in PLATFORMS and is_enabled(instrument.slug_attr)
-        ):
+class VolvoUpdateCoordinator(DataUpdateCoordinator):
+    """Volvo coordinator."""
 
-            data.instruments.add(instrument)
+    def __init__(self, hass):
+        """Initialize the data update coordinator."""
 
-            hass.async_create_task(
-                discovery.async_load_platform(
-                    hass,
-                    PLATFORMS[instrument.component],
-                    DOMAIN,
-                    (
-                        vehicle.vin,
-                        instrument.component,
-                        instrument.attr,
-                        instrument.slug_attr,
-                    ),
-                    config,
-                )
-            )
-
-    async def update(now):
-        """Update status from the online service."""
-        try:
-            if not await connection.update(journal=True):
-                _LOGGER.warning("Could not query server")
-                return False
-
-            for vehicle in connection.vehicles:
-                if vehicle.vin not in data.vehicles:
-                    discover_vehicle(vehicle)
-
-            async_dispatcher_send(hass, SIGNAL_STATE_UPDATED)
+        super().__init__(
+            hass,
+            _LOGGER,
+            name="volvooncall",
+            update_interval=DEFAULT_UPDATE_INTERVAL,
+        )
 
-            return True
-        finally:
-            async_track_point_in_utc_time(hass, update, utcnow() + interval)
+    async def _async_update_data(self):
+        """Fetch data from API endpoint."""
 
-    _LOGGER.info("Logging in to service")
-    return await update(utcnow())
+        async with async_timeout.timeout(10):
+            if not await self.hass.data[DATA_KEY].update():
+                raise UpdateFailed("Error communicating with API")
 
 
 class VolvoData:
     """Hold component state."""
 
-    def __init__(self, config):
+    def __init__(self, hass, connection: Connection, config, coordinator):
         """Initialize the component state."""
-        self.vehicles = set()
-        self.instruments = set()
+        self.hass = hass
+        self.vehicles = set()  # type: ignore[var-annotated]
```

Comment:
Type the item that the set contains to remove the type ignore.

Reply:
Unfortunately the volvooncall package doesn't expose these types

Reply:
The `vin` is a string according to how it's used in the library and here in the integration. Although there are some inconsistencies in the library in the code that looks up cached state. Eg it uses `dict.get` when it shouldn't do that, so the type looks inconsistent. But that will blow up before returning `None` as it looks.

Reply:
Was able to resolve simply by `self.vehicles: set = set()` and `self.instruments: set = set()`. Please let me know if this is not sufficient/preferred.